lib/manager/npm/extract/monorepo.ts

```diff
@@ -13,10 +22,115 @@ function matchesAnyPattern(val: string, patterns: string[]): boolean {
   return res;
 }
 
-export function detectMonorepos(
+export async function extractPnpmFilters(
+  fileName: string
+): Promise<string[] | null> {
+  try {
+    const contents = yaml.safeLoad(await readLocalFile(fileName, 'utf8'), {
+      json: true,
+    }) as any;
+    if (
+      !Array.isArray(contents?.packages) ||
+      !contents.packages.every((item) => typeof item === 'string')
+    ) {
+      logger.debug(
+        { fileName },
+        'Failed to find required "packages" array in pnpm-workspace.yaml'
+      );
+      return null;
+    }
+    return contents.packages;
+  } catch (error) {
+    logger.debug({ fileName }, 'Failed to parse pnpm-workspace.yaml');
+    return null;
+  }
+}
+
+export async function findPnpmWorkspace(
+  packageFile: string,
+  normalizedLocalDir: string
+): Promise<{ lockFilePath: string; workspaceYamlPath: string } | null> {
+  const workspaceYamlPath = await findUpInsidePath(
+    'pnpm-workspace.yaml',
+    upath.join(normalizedLocalDir, packageFile)
+  );
+
+  if (!workspaceYamlPath) {
+    logger.trace(
+      { packageFile },
+      'Failed to locate pnpm-workspace.yaml in a parent directory.'
+    );
+    return null;
+  }
+  const normalizedWorkspaceYamlPath = upath
+    .normalizeSafe(workspaceYamlPath)
+    .slice(normalizedLocalDir.length + 1);
+  const pnpmLockfilePath = getSiblingFileName(
+    normalizedWorkspaceYamlPath,
+    'pnpm-lock.yaml'
+  );
+  if (!(await localPathExists(pnpmLockfilePath))) {
+    logger.trace(
+      { normalizedWorkspaceYamlPath, packageFile },
+      'Failed to find a pnpm-lock.yaml sibling for the workspace.'
+    );
+    return null;
+  }
+  return {
+    lockFilePath: pnpmLockfilePath,
+    workspaceYamlPath: normalizedWorkspaceYamlPath,
+  };
+}
+
+export async function detectPnpm(
+  packageFiles: Partial<PackageFile>[]
+): Promise<void> {
+  logger.debug(`Detecting pnpm Workspaces`);
+  const packageFilterCache = new Map<string, string[] | null>();
+
+  const { localDir } = getAdminConfig();
```

Comment:
should `localDir` not used here, instead use `local` function from `util/fs` they will prepend all paths with `localDir` automatically.

So only use relative paths here.

Reply:
Should only use relative paths now.